numpy/core/src/umath/loops.c.src

```diff
@@ -1145,15 +1147,17 @@ TIMEDELTA_dm_m_multiply(char **args, npy_intp const *dimensions, npy_intp const
     BINARY_LOOP {
         const double in1 = *(double *)ip1;
         const npy_timedelta in2 = *(npy_timedelta *)ip2;
-        if (in2 == NPY_DATETIME_NAT) {
+        if (in2 == NPY_DATETIME_NAT || npy_isnan(in1)) {
             *((npy_timedelta *)op1) = NPY_DATETIME_NAT;
         }
         else {
-            double result = in1 * in2;
-            if (npy_isfinite(result)) {
-                *((npy_timedelta *)op1) = (npy_timedelta)result;
-            }
-            else {
+            /* `nearbyint` avoids warnings (should not matter here, though) */
+            double result = nearbyint(in1 * in2);
```

Comment:
I am not sure about this, also in that:  Is this actually enough?

The input is effectively int64, so by casting it to double we throw away precision, this means that:
```
In [6]: np.timedelta64(2**58 + 7, "us") - np.timedelta64(2**58 + 7, "us") * 1.
Out[6]: numpy.timedelta64(7,'us')
```
woops.  `long double` could fix this when available (extended precision has 64bit mantissa), but that doesn't always exist...

So, maybe the whole expectation of this rounding correct is too much, or maybe we should consider the above a serious bug?

Reply:
> So, maybe the whole expectation of this rounding correct is too much, or maybe we should consider the above a serious bug?

I think this should be considered a bug if and only if the analogous behavior in int64 is considered a bug:

```
td64 = np.timedelta64(2**58 + 7, "us")
val = td64.view("i8")

>>> val - int(val*1.0)
7
```

Reply:
Yes, it is the same thing.  But `int()` always truncates too (i.e. no correct rounding).  And you call `int()` explicitly, here we call it internally hidden to the user and thus effectively breaking the time precision for very large values.

But yeah, maybe it is just what it is, as Chuck keeps saying, maybe we really need a high-precision floating point time format (double-double, or whatever)...